# My Python Data Portfolio 2

### Aim
To take a couple of messy data sets and explore/clean them.

### About the data sets
My data sets are in the form of comma separated values (csv) and were generated by code.<br>
If you would like to know more about the creation of the data sets you can look here.

#### First, import the data sets.

In [2]:
import pandas as pd
import numpy as np
import re

measurements=pd.read_csv('measurements.csv')
active=pd.read_csv('active.csv')

#### Then look at the data of the measurements data set.<br>
- 'ID' contains the ID of the entry.
- 'Start Date' is the start date.
- 'End Date' is the end date.
- '-' is a column containing the '-' character.
- 'Measurements ' columns contain the measurements as well a date.

In [3]:
print(measurements.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100 entries, 0 to 10099
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               10085 non-null  float64
 1   Start Date       10085 non-null  object 
 2   End Date         10084 non-null  object 
 3   -                10084 non-null  object 
 4   Measurements     10082 non-null  object 
 5   Measurements 2   10085 non-null  object 
 6   Measurements 3   10084 non-null  object 
 7   Measurements 4   10081 non-null  object 
 8   Measurements 5   10084 non-null  object 
 9   Measurements 6   10089 non-null  object 
 10  Measurements 7   10089 non-null  object 
 11  Measurements 8   10088 non-null  object 
 12  Measurements 9   10084 non-null  object 
 13  Measurements 10  10085 non-null  object 
dtypes: float64(1), object(13)
memory usage: 1.1+ MB
None


#### Next, to look at the data of the active data set.<br>
- 'ID' contains the ID of the entry.
- 'Active' shows whether the entry is active or not.

In [4]:
print(active.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100 entries, 0 to 10099
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      10085 non-null  float64
 1   Active  10073 non-null  float64
dtypes: float64(2)
memory usage: 157.9 KB
None


#### Initial impressions in the 'measurements' data set.
1. The data set contains columns with different counts.<br>
This immediately shows that some cleaning will have to take place to make the counts even amongst columns.
2. The first 'Measurements' column contains no number.
3. There is a column with a title of '-' and seems to contain only '-' characters.
4. There are null values in the data set.

#### 1. Different counts in the data set.
Looking at all the unique values in the 'ID' column, there should only be one value for each ID.<br>
There are some 'ID' values that appear twice, showing a duplication.
Duplicates are removed and the code now shows that each 'ID' value has only one entry.

In [5]:
print('Measurements ', measurements['ID'].value_counts().unique())

Measurements  [2 1]


In [6]:
measurements = measurements.drop_duplicates(ignore_index=True)
print('Measurements ', measurements['ID'].value_counts().unique())

Measurements  [1]


#### 2. Measurements column with no number
Looking at the columns, the first 'Measurements' column is missing a number.
This is easy to fix by renaming the column.

In [7]:
measurements=measurements.rename(columns={'Measurements ':'Measurements 1'})
print(measurements.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9986 non-null   float64
 1   Start Date       9985 non-null   object 
 2   End Date         9984 non-null   object 
 3   -                9984 non-null   object 
 4   Measurements 1   9982 non-null   object 
 5   Measurements 2   9985 non-null   object 
 6   Measurements 3   9984 non-null   object 
 7   Measurements 4   9981 non-null   object 
 8   Measurements 5   9984 non-null   object 
 9   Measurements 6   9989 non-null   object 
 10  Measurements 7   9989 non-null   object 
 11  Measurements 8   9988 non-null   object 
 12  Measurements 9   9984 non-null   object 
 13  Measurements 10  9986 non-null   object 
dtypes: float64(1), object(13)
memory usage: 1.1+ MB
None


#### 3. Dealing with the '-' column.
Looking at the column value counts, there is only one character.<br>
The column is not needed for the data analysis and is dropped.

In [8]:
print(measurements['-'].value_counts())
measurements=measurements.drop(columns=['-'])
print(measurements.info())

-    9984
Name: -, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9986 non-null   float64
 1   Start Date       9985 non-null   object 
 2   End Date         9984 non-null   object 
 3   Measurements 1   9982 non-null   object 
 4   Measurements 2   9985 non-null   object 
 5   Measurements 3   9984 non-null   object 
 6   Measurements 4   9981 non-null   object 
 7   Measurements 5   9984 non-null   object 
 8   Measurements 6   9989 non-null   object 
 9   Measurements 7   9989 non-null   object 
 10  Measurements 8   9988 non-null   object 
 11  Measurements 9   9984 non-null   object 
 12  Measurements 10  9986 non-null   object 
dtypes: float64(1), object(12)
memory usage: 1015.8+ KB
None


#### 4. Null values
Looking at the values and null values in the data set <br>
- There are 10000 entries.
- Every column has values missing.
- There are
    - 9808 rows with no missing values.
    - 191 rows with one value missing.
    - 1 row with two values missing.
- About 2% of the rows have at least one missing value.

In [9]:
print(measurements.info())
print(measurements.shape[0])
missing_values = measurements.isna()
print(missing_values.sum(axis=1).value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9986 non-null   float64
 1   Start Date       9985 non-null   object 
 2   End Date         9984 non-null   object 
 3   Measurements 1   9982 non-null   object 
 4   Measurements 2   9985 non-null   object 
 5   Measurements 3   9984 non-null   object 
 6   Measurements 4   9981 non-null   object 
 7   Measurements 5   9984 non-null   object 
 8   Measurements 6   9989 non-null   object 
 9   Measurements 7   9989 non-null   object 
 10  Measurements 8   9988 non-null   object 
 11  Measurements 9   9984 non-null   object 
 12  Measurements 10  9986 non-null   object 
dtypes: float64(1), object(12)
memory usage: 1015.8+ KB
None
10000
0    9808
1     191
2       1
dtype: int64


2% of rows missing values is not a very large amount.<br>
However, it is worth checking if any 192 rows can be salvaged.<br>
Looking at a row from the data set, most columns have unique entries.<br>
The 'Start Date' and 'End Date' columns have many repeat values.

In [10]:
print(measurements.iloc[0])
for i in measurements.columns:
    print(measurements[i].value_counts().unique())

ID                                                              8869
Start Date                                                2019-04-23
End Date                                                  2020-04-22
Measurements 1     Value of 122.92935410861622 recorded on 2019-0...
Measurements 2     Value of 169744.45974638883 recorded on 2019-1...
Measurements 3     Value of 161.7788486722056 recorded on 2019-12-18
Measurements 4     Value of 1.096654390606948 recorded on 2020-02-08
Measurements 5     Value of 17765.89301205467 recorded on 2020-02-19
Measurements 6     Value of 9.717273558807952 recorded on 2020-02-22
Measurements 7     Value of 11075.127874101461 recorded on 2020-0...
Measurements 8     Value of 202.49040954654967 recorded on 2020-0...
Measurements 9     Value of 19443.38954424761 recorded on 2020-04-01
Measurements 10    Value of 16.142915057513967 recorded on 2020-0...
Name: 0, dtype: object
[1]
[9985]
[9984]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]


Looking at the 'Start Date' and 'End Date' columns.
- About 99% of all rows contain values.
- The rows that are not null contain a single value for each column.

In [11]:
print(measurements['Start Date'].value_counts())
print(measurements['End Date'].value_counts())

2019-04-23    9985
Name: Start Date, dtype: int64
2020-04-22    9984
Name: End Date, dtype: int64


Filling null values in start and end date columns.

In [12]:
start_date=measurements['Start Date'].value_counts().index[0]
end_date=measurements['End Date'].value_counts().index[0]
measurements['Start Date']=start_date
measurements['End Date']=end_date

print(measurements['Start Date'].value_counts())
print(measurements['End Date'].value_counts())

2019-04-23    10000
Name: Start Date, dtype: int64
2020-04-22    10000
Name: End Date, dtype: int64


Looking at the measurements data set again<br>
- There are 10000 entries.
- Two columns do not have missing values.
- There are
    - 9839 rows with no missing values.
    - 160 rows with one value missing.
    - 1 row with two values missing.
- About 2% of the rows have at least one missing value.
- 31 rows were salvaged.

In [13]:
print(measurements.info())
print(measurements.shape[0])
missing_values = measurements.isna()
print(missing_values.sum(axis=1).value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9986 non-null   float64
 1   Start Date       10000 non-null  object 
 2   End Date         10000 non-null  object 
 3   Measurements 1   9982 non-null   object 
 4   Measurements 2   9985 non-null   object 
 5   Measurements 3   9984 non-null   object 
 6   Measurements 4   9981 non-null   object 
 7   Measurements 5   9984 non-null   object 
 8   Measurements 6   9989 non-null   object 
 9   Measurements 7   9989 non-null   object 
 10  Measurements 8   9988 non-null   object 
 11  Measurements 9   9984 non-null   object 
 12  Measurements 10  9986 non-null   object 
dtypes: float64(1), object(12)
memory usage: 1015.8+ KB
None
10000
0    9839
1     160
2       1
dtype: int64


Any row with a missing value makes the row skew the results of analysis.<br>
The 161 rows with at least one missing value will be dropped.<br>
Over 98% (9839) of the rows remain.<br>
The primary data set now contains columns of equal length.

In [14]:
measurements=measurements.dropna().reset_index(drop=True)
print(measurements.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9839 entries, 0 to 9838
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9839 non-null   float64
 1   Start Date       9839 non-null   object 
 2   End Date         9839 non-null   object 
 3   Measurements 1   9839 non-null   object 
 4   Measurements 2   9839 non-null   object 
 5   Measurements 3   9839 non-null   object 
 6   Measurements 4   9839 non-null   object 
 7   Measurements 5   9839 non-null   object 
 8   Measurements 6   9839 non-null   object 
 9   Measurements 7   9839 non-null   object 
 10  Measurements 8   9839 non-null   object 
 11  Measurements 9   9839 non-null   object 
 12  Measurements 10  9839 non-null   object 
dtypes: float64(1), object(12)
memory usage: 999.4+ KB
None


#### Initial impressions in the 'active' data set.
1. The data set contains columns with different counts.<br>
This immediately shows that some cleaning will have to take place to make the counts even amongst columns.
2. There are null values in the data set.


#### 1. Different counts in the data set.
Looking at all the unique values in the 'ID' column, there should only be one value for each ID.<br>
There are some 'ID' values that appear twice, showing a duplication.
Duplicates are removed and the code now shows that each 'ID' value has only one entry.

In [15]:
print('Active ', active['ID'].value_counts().unique())

Active  [2 1]


Duplicate rows are removed, leaving 9988 rows remaining.<br>
Looking at the columns there are different numbers of entries.<br>


In [16]:
active = active.drop_duplicates(ignore_index=True)
print('Active ', active['ID'].value_counts().unique())
print(active.info())

Active  [1]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9988 entries, 0 to 9987
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      9986 non-null   float64
 1   Active  9962 non-null   float64
dtypes: float64(2)
memory usage: 156.2 KB
None


#### 2. Null values in the data set.
Looking at the first five rows of the data set
- The 'ID' is unordered, so missing values would not be easy to salvage.<br>
- The 'Active' value shows whether or not the ID is active.  This again would be hard to salvage.

Assuming missing 'ID' values and missing 'Active' values are not on the same row, then this leaves a maximum of 2+26=28 rows to be removed.<br>
This would leave 9960 rows remaining (about 99.7%).

In [17]:
print(active.head(10))
print((9988-(2+26))/9988)

       ID  Active
0  8869.0     1.0
1  6522.0     1.0
2  3109.0     1.0
3  9434.0     1.0
4  8073.0     1.0
5   731.0     1.0
6  2409.0     1.0
7  2507.0     0.0
8  6985.0     0.0
9   791.0     1.0
0.9971966359631558


Rows with null values are dropped, leaving 9960 rows remaining.

In [18]:
active=active.dropna().reset_index(drop=True)
print(active.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9960 entries, 0 to 9959
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      9960 non-null   float64
 1   Active  9960 non-null   float64
dtypes: float64(2)
memory usage: 155.8 KB
None


#### Joining the two data sets
Now that the individual data sets have been clean up, they can now be joined together.<br>
The two data sets will be joined on the 'ID' column.<br>
After the join, the primary data set 'measurements' will contain an extra column 'Active'.<br>
The expanded data set now contains some null values that need to be removed.<br>

In [19]:
measurements=measurements.join(active.set_index('ID'), on='ID')
print(measurements.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9839 entries, 0 to 9838
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9839 non-null   float64
 1   Start Date       9839 non-null   object 
 2   End Date         9839 non-null   object 
 3   Measurements 1   9839 non-null   object 
 4   Measurements 2   9839 non-null   object 
 5   Measurements 3   9839 non-null   object 
 6   Measurements 4   9839 non-null   object 
 7   Measurements 5   9839 non-null   object 
 8   Measurements 6   9839 non-null   object 
 9   Measurements 7   9839 non-null   object 
 10  Measurements 8   9839 non-null   object 
 11  Measurements 9   9839 non-null   object 
 12  Measurements 10  9839 non-null   object 
 13  Active           9813 non-null   float64
dtypes: float64(2), object(12)
memory usage: 1.1+ MB
None


Dropping rows with null values has removed an extra 26 rows.
There are now 9813 rows remaining. <br>
This means a loss of about
- 2.9% of the original data set.
- 1.9% of the primary data set after dropping duplicates.
- 0.3% of the primary data set after dropping duplicates and null containing rows.

The relevance of this depends on the tolerance to dropping rows, the importance of duplicates and the importance of the rows which contained missing values.<br>
I believe that a less than 3% loss from the raw data set is acceptable in this instance.

In [20]:
measurements=measurements.dropna().reset_index(drop=True)
print(measurements.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               9813 non-null   float64
 1   Start Date       9813 non-null   object 
 2   End Date         9813 non-null   object 
 3   Measurements 1   9813 non-null   object 
 4   Measurements 2   9813 non-null   object 
 5   Measurements 3   9813 non-null   object 
 6   Measurements 4   9813 non-null   object 
 7   Measurements 5   9813 non-null   object 
 8   Measurements 6   9813 non-null   object 
 9   Measurements 7   9813 non-null   object 
 10  Measurements 8   9813 non-null   object 
 11  Measurements 9   9813 non-null   object 
 12  Measurements 10  9813 non-null   object 
 13  Active           9813 non-null   float64
dtypes: float64(2), object(12)
memory usage: 1.0+ MB
None


#### Sorting the values
Up until now, the 'ID' column has been unordered.
Now that the data sets have been cleaned and merged, the set is now sorted by 'ID'.<br>

In [21]:
print(measurements['ID'].head())
measurements=measurements.sort_values(by='ID').reset_index(drop=True)
print(measurements['ID'].head())

0    8869.0
1    6522.0
2    3109.0
3    9434.0
4    8073.0
Name: ID, dtype: float64
0    1.0
1    2.0
2    3.0
3    4.0
4    5.0
Name: ID, dtype: float64


#### Extracting values from string
Looking at the first row in all the 'Measurements' columns, there are two values stored together as a string.<br>
- The measurement value.
- The date the measurement was taken.
- The data is stored in the format
    - 'Value of (measurement) recorded on (date)'

In [22]:
for i in range(10):
    print(measurements['Measurements {}'.format(i+1)][0])

Value of 122.674488292129 recorded on 2019-05-28
Value of 125876.73652156458 recorded on 2019-07-11
Value of 128.28315824771258 recorded on 2019-11-15
Value of 0.9507125385895309 recorded on 2019-11-16
Value of 18685.103450216673 recorded on 2019-12-15
Value of 9.899248175980052 recorded on 2019-12-24
Value of 10535.694060169368 recorded on 2020-02-02
Value of 193.89724334859324 recorded on 2020-02-03
Value of 15091.208538795898 recorded on 2020-03-01
Value of 15.37298207359299 recorded on 2020-04-04


Splitting a measurement string when there is a space gives
0. 'Value'
1. 'of'
2. (measurement value)
3. 'recorded'
4. 'on'
5. (date value)

In [23]:
print(measurements['Measurements 1'].head().str.split())

0    [Value, of, 122.674488292129, recorded, on, 20...
1    [Value, of, 126.0508399952744, recorded, on, 2...
2    [Value, of, 127.67031994040427, recorded, on, ...
3    [Value, of, 124.74240413560719, recorded, on, ...
4    [Value, of, 123.76310913073245, recorded, on, ...
Name: Measurements 1, dtype: object


For each 'Measurements' column
    - Split the measurement string, creating a list of values
    - Assign the element containing the date value to a new column ('Measurement Date')
    - Assign the element containing the measurement value to the measurement column ('Measurements')
The data set now contains ten new columns, one for each measurement column.<br>
The 'Measurement Date' columns contain the date of the measurement.<br>
The 'Measurements' columns now contain just the measurement value.

In [24]:
for i in range(10):
    measurements['Measurement Date {}'.format(i+1)]=measurements['Measurements {}'.format(i+1)].apply(lambda x: x.split()[5])
    measurements['Measurements {}'.format(i+1)]=measurements['Measurements {}'.format(i+1)].apply(lambda x: x.split()[2])
print(measurements.iloc[0])

ID                                      1
Start Date                     2019-04-23
End Date                       2020-04-22
Measurements 1           122.674488292129
Measurements 2         125876.73652156458
Measurements 3         128.28315824771258
Measurements 4         0.9507125385895309
Measurements 5         18685.103450216673
Measurements 6          9.899248175980052
Measurements 7         10535.694060169368
Measurements 8         193.89724334859324
Measurements 9         15091.208538795898
Measurements 10         15.37298207359299
Active                                  1
Measurement Date 1             2019-05-28
Measurement Date 2             2019-07-11
Measurement Date 3             2019-11-15
Measurement Date 4             2019-11-16
Measurement Date 5             2019-12-15
Measurement Date 6             2019-12-24
Measurement Date 7             2020-02-02
Measurement Date 8             2020-02-03
Measurement Date 9             2020-03-01
Measurement Date 10            202

#### Changing the data types
The size of the data set is 1.8MB+<br>
Looking at the data set<br>
    - 'ID' could be stored as an integer.
    - 'Start Date', 'End Date' and 'Measurement Date' could be stored as a date object.
    - 'Active' could be stored as a boolean.
    - 'Measurements' could be stored as float.

In [25]:
print(measurements.info())

measurements['ID']=measurements['ID'].astype('int')
measurements['Start Date']=measurements['Start Date'].astype('datetime64')
measurements['End Date']=measurements['End Date'].astype('datetime64')
measurements['Active']=measurements['Active'].astype('boolean')
for i in range(10):
    measurements['Measurements {}'.format(i+1)]=measurements['Measurements {}'.format(i+1)].astype('float')
    measurements['Measurement Date {}'.format(i+1)]=measurements['Measurement Date {}'.format(i+1)].astype('datetime64')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   9813 non-null   float64
 1   Start Date           9813 non-null   object 
 2   End Date             9813 non-null   object 
 3   Measurements 1       9813 non-null   object 
 4   Measurements 2       9813 non-null   object 
 5   Measurements 3       9813 non-null   object 
 6   Measurements 4       9813 non-null   object 
 7   Measurements 5       9813 non-null   object 
 8   Measurements 6       9813 non-null   object 
 9   Measurements 7       9813 non-null   object 
 10  Measurements 8       9813 non-null   object 
 11  Measurements 9       9813 non-null   object 
 12  Measurements 10      9813 non-null   object 
 13  Active               9813 non-null   float64
 14  Measurement Date 1   9813 non-null   object 
 15  Measurement Date 2   9813 non-null   o

All the data types have been changed, reducing the size of the data set to 1.7MB.<br>
This is a reduction of about 5.5%.

In [26]:
print(measurements.info())
print(measurements.iloc[0])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID                   9813 non-null   int32         
 1   Start Date           9813 non-null   datetime64[ns]
 2   End Date             9813 non-null   datetime64[ns]
 3   Measurements 1       9813 non-null   float64       
 4   Measurements 2       9813 non-null   float64       
 5   Measurements 3       9813 non-null   float64       
 6   Measurements 4       9813 non-null   float64       
 7   Measurements 5       9813 non-null   float64       
 8   Measurements 6       9813 non-null   float64       
 9   Measurements 7       9813 non-null   float64       
 10  Measurements 8       9813 non-null   float64       
 11  Measurements 9       9813 non-null   float64       
 12  Measurements 10      9813 non-null   float64       
 13  Active               9813 non-nul

The primary data set is now exported to csv.

In [27]:
measurements.to_csv('measurements_and_active.csv', index=False)

#### Conclusion
The two data sets have been explored, cleaned and had their duplicates and null value rows removed.<br>
The losses have been calculated and found to be acceptable.<br>
The two sets were merged into a more manageable single data set.<br>
The date and measurement values were extracted from the measurement string.<br>
The data types have been changed, reflecting the data better and reducing the data set size in MB.

